bolero-generator/src/arbitrary.rs

```diff
@@ -0,0 +1,70 @@
+use core::{cmp, marker::PhantomData};
+
+use arbitrary::{Arbitrary, Unstructured};
+
+use crate::{gen_with, Driver, ValueGenerator};
+
+pub struct ArbitraryGenerator<T>(PhantomData<T>);
+
+const ABUSIVE_SIZE: usize = 1024 * 1024;
+const MIN_INCREASE: usize = 32;
+
+impl<T> ValueGenerator for ArbitraryGenerator<T>
+where
+    T: for<'a> Arbitrary<'a>,
+{
+    type Output = T;
+
+    fn generate<D: Driver>(&self, driver: &mut D) -> Option<T> {
+        let size = T::size_hint(0);
+        let mut data = match T::size_hint(0) {
+            (min, Some(max)) if max < ABUSIVE_SIZE => gen_with::<Vec<u8>>()
+                .len(min..=max)
+                .generate(&mut *driver)?,
+            (min, _) => gen_with::<Vec<u8>>().len(min).generate(&mut *driver)?,
+        };
+        loop {
+            match Unstructured::new(&data).arbitrary() {
+                Ok(res) => return Some(res),
+                Err(arbitrary::Error::NotEnoughData) => (), // fall-through to another iter
```

Comment:
We should only loop if the `DriverMode` is `Forced`. The `Direct` mode works a bit better for coverage-guided fuzzers since you'd signaling that the input wasn't structured how the `Arbitrary` generator wanted it.